### ISSUE TITLE
Error nil on retryFetch Function

### ISSUE THREAD

**Issue by muh-zaki** [amplitude/experiment-go-server#1]
Hi,

## Context
I use this library in my golang server to fetch data from Amplitude. While testing retry, I realized enabling `FetchRetries` will always give error output of `client.Fetch` function as nil, even though there is an error.

I simulate error by setting `FetchTimeout: 1 nanosecond`
Below is the log for testing without using retry by setting `FetchRetries: 0`

```
func testFunc() {
	...
	client := remote.Initialize(_APIKEY_, &remote.Config{
		Debug:        true,
		FetchTimeout: 1 * time.Nanosecond,
		RetryBackoff: &remote.RetryBackoff{
			FetchRetries:            0,
			FetchRetryBackoffMin:    100 * time.Millisecond,
			FetchRetryBackoffMax:    1 * time.Second,
			FetchRetryBackoffScalar: 2,
		},
	})
	variants, err := client.Fetch(__user__)
	if err != nil {
		fmt.Println("Error fetch: ", err)

		if err, ok := (err).(net.Error); ok && err.Timeout() {
			fmt.Println("Error timeout: ", err)
		}
		return
	}
	payload = variants[__key__].Payload
	if payload == nil {
		fmt.Println("empty payload")
	}
	...
}
```

<img width="1327" alt="Screen Shot 2022-06-20 at 20 36 12" src="https://user-images.githubusercontent.com/34985453/174614671-f648f82d-e0ed-4291-bbbf-1a9030034bb7.png">

From the result, we can see that the timeout error was returned by `Fetch` function.

## Current Behavior
On the other hand, when I set `FetchRetries: 5`, the timeout error does not returned by `Fetch` function, instead, it returned `nil` error with empty result. We can see that on the log below.

<img width="1320" alt="Screen Shot 2022-06-20 at 20 49 09" src="https://user-images.githubusercontent.com/34985453/174619008-51bf229c-8794-4f27-a3b5-f00ff950a986.png">

See the red boxes. Every retries gave timeout error, but the conclusion have the error as `<nil>` that returned as `Fetch` results. This issue make it hard to handle error separately and track the error code for alert or debugging.

## Expected Behavior
Even in retryFetch, the `Fetch` function should return the error, at least error from the last retry.

## Possible Solution

Deep-dive further to the code, in the `client.go` file. I think the issue happen because of the code in yellow block. Instead of set the value of `err` variable that created in `line 97`, the code in line `104` create new local variable (scoped inside `for` block). Hence, the error that printed and returned in `lines 115-116` is always be `nil`.

<img width="979" alt="Screen Shot 2022-06-20 at 21 07 38" src="https://user-images.githubusercontent.com/34985453/174620853-4fa39b46-e12a-402b-ad23-1e4a34045186.png">

I think the issue can be solved by declaring the variables `variants` and `err` before the `for` loop, then just use assignment without initialization when call `doFetch` function.
<img width="905" alt="Screen Shot 2022-06-20 at 21 18 31" src="https://user-images.githubusercontent.com/34985453/174622750-7fd11c22-4f1f-4605-b44c-8dbd0549a2b2.png">

## Question 
Is this an expected behavior for `retryFetch` function?

## Other
Go version: 1.17

Thank you.

**Comment by bgiori** [comment]
@muh-zaki This issue has been fixed in `v0.4.6` 

Thanks so much for the detailed ticket! You definitely saved me some time.

**Comment by muh-zaki** [comment]
Hi @bgiori ,
No problem! I have tested the `v0.4.6` and it is working perfectly.
Thank you!

<img width="1252" alt="Screen Shot 2022-06-21 at 10 10 08" src="https://user-images.githubusercontent.com/34985453/174708442-8a036e79-3df1-4f43-afb2-955b6101ec87.png">
